### PR TITLE
Icon Smoothing 

### PR DESCRIPTION
**Pull request by xxalpha**
For all your smoothing needs

 - One ~~datum~~ method to smooth them all
 - Everything that should icon smooth uses this
 - Removed grass icon smoothing

All sprite work done by @WJohn also doing this was his idea.

Now please tell me how my code is shit, I don't care anymore. :speedboat:

Fixes https://github.com/tgstation/-tg-station/issues/10320 
Fixes https://github.com/tgstation/-tg-station/issues/10243 
Fixes https://github.com/tgstation/-tg-station/issues/10240 
Fixes https://github.com/tgstation/-tg-station/issues/10140

### CODE REVIEW

**Comment by WJohn** [comment]
:+1: x :100: 

Actually surprised at how quickly this things got resolved.

File: code/game/objects/items/weapons/storage/fancy.dm [hunk 1]
```diff
@@ -25,6 +25,20 @@
 	..()
 	for(var/i = 1 to storage_slots)
 		new spawn_type(src)
+	world << "NORTH ([NORTH])"
```
Comment (optimumtact) [inline code/game/objects/items/weapons/storage/fancy.dm]: Left in debug?

**Comment by Razharas** [comment]
Two issues already:
1 - No diagonal cases? See diagonal walls and L shape tables to get the issue
2 - N-datums = N-atoms to smooth? And each datum has a separate set of images? If yes than horrible

File: tgstation.dme [hunk 2]
```diff
@@ -263,7 +264,6 @@
 #include "code\game\say.dm"
 #include "code\game\shuttle_engines.dm"
 #include "code\game\skincmd.dm"
-#include "code\game\smoothwall.dm"
```
Comment (optimumtact) [inline tgstation.dme]: did you mean to untick smoothwall?

Comment (xxalpha) [inline tgstation.dme]: I meant to delete it entirely, but I forgot.

**Comment by xxalpha** [comment]
> diagonal walls

Are you drunk

**Comment by optimumtact** [comment]
I think this is a good change :+1:

File: code/datums/smooth.dm [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+		else
+			sdir = "nw"
+	else
+		if(cardinal_adj & NORTH)
+			sdir = "n"
+		else if(cardinal_adj & WEST)
+			sdir = "w"
+		else
+			sdir = "i"
+	top_left = image(holder.icon, "1-[sdir]")
```
Comment (optimumtact) [inline code/datums/smooth.dm]: This makes me nervous, are you able to make images once, store them in a shared global list then look them up to see if they've been created before and reuse that image? You'd need a key system like objtype-icon-sdir or some such

Creating lots of images is bad for performance (especially where a lot of the images are reused)

Comment (xxalpha) [inline code/datums/smooth.dm]: Yes it is and yes I will try that.

File: code/game/objects/structures/window.dm [hunk 4]
```diff
@@ -393,20 +392,33 @@
 		if(!src) return
 		if(!fulltile)
 			return
-		var/junction = 0 //will be used to determine from which side the window is connected to other windows
+		/*var/junction = 0 //will be used to determine from which side the window is connected to other windows
```
Comment (optimumtact) [inline code/game/objects/structures/window.dm]: Just chop this code out if it's unused

File: code/game/objects/structures/window.dm [hunk 5]
```diff
@@ -393,20 +392,33 @@
...
 
-		overlays.Cut()
 		var/ratio = health / maxhealth
 		ratio = Ceiling(ratio*4) * 25
-		overlays += "damage[ratio]"
-		return
+
+		if(smoother)
+			smoother.smooth()
+			sleep(2)
```
Comment (optimumtact) [inline code/game/objects/structures/window.dm]: why is there a sleep here?

Comment (xxalpha) [inline code/game/objects/structures/window.dm]: Good question

File: code/datums/smooth.dm (hunk 3 above)
Comment (optimumtact) [inline code/datums/smooth.dm]: I've just realised you can generate the images at the time they are needed once
i.e the flow is
check if image stored in global list:
if yes - use
else - generate and store in global list

Comment (Incoming5643) [inline code/datums/smooth.dm]: Female outfit generation uses a one time on demand image generation like what you're talking about here.

File: code/datums/smooth.dm [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+		for(var/direction in cardinal)
+			if(find_type_in_direction(holder, direction, siblings))
+				cardinal_adj |= direction
+		for(var/direction in diagonals)
+			if(find_type_in_direction(holder, direction, siblings))
+				diagonal_adj |= direction
+
+/datum/tile_smoother/proc/smooth()
+	if(!enabled)	return
+	spawn(2)
```
Comment (MrPerson) [inline code/datums/smooth.dm]: This spawn() is completely unnecessary

File: code/datums/smooth.dm [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+				var/turf/T = locate(source.x + x_offset, source.y + y_offset, source.z)
+				if(T.type == a_type)
+					return T
+			else
+				var/atom/A = locate(a_type) in locate(source.x + x_offset, source.y + y_offset, source.z)
+				if(A && A.type == a_type)
+					return A
+		return null
+	var/atom/A = locate(source.type) in locate(source.x + x_offset, source.y + y_offset, source.z)
+	return A && A.type == source.type ? A : null
```
Comment (MrPerson) [inline code/datums/smooth.dm]: I'm not saying you should change this, but you could get the same effect with return A && A.type == source.type && A

File: code/datums/smooth.dm (hunk 6 above)
Comment (optimumtact) [inline code/datums/smooth.dm]: I suspect it was related to the singulo causing lots of resmoothing that would then be destroyed the next second

File: code/game/objects/structures/window.dm (hunk 5 above)
Comment (optimumtact) [inline code/game/objects/structures/window.dm]: I suspect it was related to the singulo causing lots of resmoothing that would then be destroyed the next second

File: code/game/objects/items/weapons/storage/fancy.dm (hunk 1 above)
Comment (xxalpha) [inline code/game/objects/items/weapons/storage/fancy.dm]: Holyshit i forgot all about these

File: code/datums/smooth.dm (hunk 7 above)
Comment (xxalpha) [inline code/datums/smooth.dm]: I need to be returning atoms here, actually.

**Comment by freerealestate** [comment]
Does this fix that thing where if you break a wall the connecting walls retain their shape instead of magically morphing to look standalone? Because I like that.

**Comment by xxalpha** [comment]
Yes, it updates the walls.

**Comment by xxalpha** [comment]
It refuses to compile with metastation and ministation. I will give up trying.

**Comment by xxalpha** [comment]
Okay this doesn't actually fix https://github.com/tgstation/-tg-station/issues/10140 because I didn't make those walls smooth. Is it worth it making the indestructible z2 walls smooth?

**Comment by xxalpha** [comment]
Don't merge this just yet.

File: code/__HELPERS/icon_smoothing.dm [hunk 8]
```diff
@@ -0,0 +1,192 @@
+/atom/var/smooth = 0
+/atom/var/top_left_corner
+/atom/var/top_right_corner
+/atom/var/bottom_left_corner
```
Comment (optimumtact) [inline code/__HELPERS/icon_smoothing.dm]: Thanks @Razharas

**Comment by xxalpha** [comment]
I am done with this.

**Comment by Rodocks** [comment]
how is this better than the old system?
instead of having predefined sprites for every direction you have 16x16 sprites it puts more stress on the server and the poor people that have to sprite for this

**Comment by xxalpha** [comment]
Smaller file sizes.

**Pull request merged**